[1단계 - 출석 구현] 듀이(이주희) 미션 제출합니다.

ljhee92 opened the pull request:
안녕하세요 샐리, 7기 BE 크루 듀이입니다.
먼저 소중한 시간 들여 리뷰해주셔서 감사합니다!
1단계는 페어와 함께 ‘돌아가는 쓰레기’라도 만드는 것을 최우선 목표로 했기 때문에 리팩토링을 충분히 하지 못했습니다.
보시기에 코드가 많이 정신없고 복잡할 수 있으니 미리 양해 부탁드려요🥹🥹

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - https://prolog.techcourse.co.kr/studylogs/3898


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

- 모든 메서드의 indent를 1로 유지하지 못했습니다.
- 모든 원시 값과 문자열을 포장하지 못했습니다.
- 모든 엔티티를 작게 유지하지 못했다고 생각합니다.
- getter를 사용했습니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

### 도메인 분리와 범위
- 도메인을 어떻게 분리하면 효과적일지 궁금합니다.
이번 출석 미션에서는 ‘출석’이기 때문에 ‘출석’ 도메인을 먼저 분리하고 출석에서부터 필요한 것들을 분리해나갔습니다.
그런데 구현해 나갈 수록 출석이 가장 중요한 도메인이긴 하지만 가장 작은 도메인은 아닌 것 같아 TDD를 놓치게 된 것 같습니다.
샐리는 어떤 방식으로 도메인을 분리해나가시는지 궁금합니다.

- 도메인의 범위가 어디까지여야 할지 궁금합니다.
출석에서 크루, 출석 상태, 제적 위험을 나누는 페널티까지 분리를 했는데, 가장 작은 단위로 따지게 된다면 등교일, 운영 시간도 나누어야 하지 않을까 하는 생각도 들었습니다.
리팩토링을 충분히 하지 못했다고 생각이 드는 것도 이런 이유 때문이기도 합니다.

### 공휴일 Enum 처리
- 처음 구현할 때에는 요구사항에 따라 2024년 12월 기준으로 구현했기 때문에 공휴일을 우선 ``LocalDate.of(2024, 12, 25)``로 직접 주입해주었습니다.
그런데 시스템 시간을 현재로 변경하고, 2025년 공휴일로 맞추려고 하니 한 번에 관리하고자 Enum을 도입하여 상수화시키게 되었는데요.
만약 공휴일이 추가되거나 하면 공휴일 Enum을 변경해야 할텐데 Enum 외에 방법이 있을지 궁금합니다.

### view의 책임
- 레벨1 미션에서 MVC 패턴으로 리팩토링할 때, view는 domain에 의존할 수 있었기 때문에 이번에도 view에서는 domain 패키지의 클래스들을 주입받아 출력 형식에 맞추어 정렬했습니다.
그런데 이렇게 하니 domain의 내부 값들에 따라 정렬을 하는 것이 view의 책임이 맞는가?하는 생각도 들었습니다.
예를 들어 제적 위험자를 확인하는 기능에서 아래와 같은 요구사항이 있었습니다.
  - 제적 위험자는 제적 대상자, 면담 대상자, 경고 대상자순으로 출력하며, 대상 항목별 정렬 순서는 지각을 결석으로 간주하여 내림차순한다. 출석 상태가 같으면 닉네임으로 오름차순 정렬한다.
- 저와 페어는 view에서 해당 정렬 기능을 책임지도록 했는데, 이게 정말 view의 책임이 맞을지 샐리의 생각도 궁금합니다.

### nested record
- 저는 한 번도 사용해본 적이 없었는데, 파일을 파싱할 때와 제적 위험자 출력 시 페어가 nested record를 사용하는 것이 어떠냐고 제안해서 사용하게 되었습니다. 구글링 해보니 참고할 만한 자료들이 딱히 나오지는 않는데, 사용해도 괜찮을지, 혹시 추천해주실만한 참고 자료가 있을지 궁금합니다.

Minuring committed:
docs: 기능 요구사항 정리

Minuring committed:
feat: 크루 객체 생성

닉네임을 가지는 크루입니다.

Minuring committed:
feat: 출석 객체 생성

닉네임과 등교 시간으로 출석 객체를 생성한다.
등교 날짜가 주말이면 예외가 발생한다.

Minuring committed:
chore: domain 테스트 패키지 분류

Minuring committed:
feat: 출석 상태 기능 추가

요일 별 교육 운영시간에 따른 출석 상태를 결정합니다.

Minuring committed:
feat: 공휴일 출석 예외처리 기능

Minuring committed:
feat: 출석 수정 기능

Minuring committed:
feat: 출석 기록 파일 입력

Minuring committed:
refactor: 출석 객체에서 크루 제거

출석 객체에서 크루가 사용되지 않고 크루 별 출석을 더 쉽게 관리하기 위해서 제거합니다.

Minuring committed:
feat: 크루별 출석 기록 기능

크루별 출석을 저장하고 꺼내오는 기능입니다.
기능 추가에 따라 기존의 파일 입력 기능을 수정했습니다.

Minuring committed:
feat: 크루 출석 기록 확인 시 결석 여부 확인 기능

Attendances와 Attendance 공통적으로 사용되는 주말 확인 로직을 유틸로 분리합니다.

Minuring committed:
feat: 제적 위험자 확인 기능

지각 횟수와 결석 횟수로 제적 여부를 확인합니다.

Minuring committed:
feat: 출석 확인 기능 입출력

최상위 패키지 attendance를 추가합니다.

Minuring committed:
feat: 출석 수정 기능 입출력

Minuring committed:
feat: 크루별 출석 기록 확인 기능 입출력

Minuring committed:
feat: 제적 위험자 확인 기능 입출력

기능 확인을 위해 csv파일을 수정했습니다.

Minuring committed:
feat: 기능 선택 화면 입출력

Minuring committed:
refactor: 시스템 시간 변경

Minuring committed:
refactor: 오류 메시지 출력

Minuring committed:
refactor: 가독성 향상 및 공휴일 상수화

- 메서드 분리
- 변수명 변경
- 공휴일 enum으로 분리

Minuring committed:
refactor: 날짜 포맷팅 한국어 적용

dusdn1702 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,117 @@
+    private Crews crews;
+
+    public AttendanceController() throws IOException {
+        String path = "src/main/resources/attendances.csv";
+        FileContents fileContents = AttendanceFileReader.read(path);
```
요걸 controller에서 수행하신 이유가 궁금해요!

ljhee92 replied:
큰 의미없이 파일의 정보를 프로그램 실행 시점에 가져와야겠다고 생각해서 controller가 생성될 때 수행해주었습니다.
그런데 지금 생각해보니 프로그램의 실행 시점이라면 Application 에서 읽은 뒤 controller에 주입해주는 것이 더 알맞다고 생각되네요!

ljhee92 replied:
테스트 코드와 프로덕션 코드에서 사용하는 csv 파일 경로를 다르게 주입해주기 위해 controller에서 파일 경로를 선언하고 ``AttendanceFileReader``로부터 초기 데이터를 읽어오도록 했습니다..!

dusdn1702 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,117 @@
+        crews = fileContents.crews();
+    }
+
+    public void run() {
```
전체적으로 기능이 잘 수행되지 않는게 있는 것 같아요! 확인 부탁드릴게요~
3.4. 확인 시 java.lang.IndexOutOfBoundsException: Index 13 out of bounds for length 13 가 발생하네요 👀

ljhee92 replied:
아앗..😨 꼼꼼하게 확인하지 못해 죄송합니다🥹 확인하여 반영해두었습니다!

dusdn1702 commented on src/main/java/attendance/controller/AttendanceController.java:
```diff
@@ -0,0 +1,117 @@
+
+    private LocalTime getModifyingCheckinTime() {
+        String inputModifyingCheckinTime = InputView.readModifyingCheckinTime();
+        return LocalTime.parse(inputModifyingCheckinTime);
```
이 부분을 controller에서 수행해주신 이유가 궁금해요!

ljhee92 replied:
view는 사용자의 입력을 그대로 넘겨줄 뿐 파싱할 책임은 없다고 생각했고, domain도 필요한 데이터를 받아 비즈니스 로직만 처리해야 한다고 생각했기 때문입니다. controller는 view와 domain을 연결해주는 역할이기 때문에 domain이 필요로 하는 데이터를 view에서 받은 입력에서부터 파싱하는 것 또한 controller에게 부여했습니다.

dusdn1702 commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+
+    private void setStatus(LocalDateTime dateTime) {
```
이름이 setter와 겹치는데 같은 역할을 아니라 다른 메서드명을 가져보는건 어떨까요?

ljhee92 replied:
네 이부분도 메서드명 고민하여 반영해두었습니다!

dusdn1702 commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    private void determineStatus(LocalTime time, int hour) {
+        if ((time.isAfter(LocalTime.of(8, 0)) && time.isBefore(LocalTime.of(hour, 5)))
```
어떤 값인지 알기 어려워 상수화 하면 좋을 것 같아요!

ljhee92 replied:
넵 가독성 고려하여 상수처리 해두었습니다😄

dusdn1702 commented on src/main/java/attendance/domain/Attendance.java:
```diff
@@ -0,0 +1,76 @@
+
+public class Attendance {
+
+    private LocalDateTime dateTime;
```
DateTime이라는 예약어보다 좋은 변수명이 있을 것 같아요!

ljhee92 replied:
네 확인하여 반영했습니다!

dusdn1702 commented on src/main/java/attendance/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,18 @@
+    LATE("지각")
+    ;
+
+    private String koreanName;
```
출석 상태의 한국이름이란 이름이 어색하게 느껴지네요..ㅎㅎ

ljhee92 replied:
Enum 요소가 영어로 되어 있어 한국어 이름으로 명명하였는데, 출석 상태의 한국어 이름이라고 하니 뭔가 어색하네요😅 ``Penalty`` Enum도 함께 ``status``로 변경해두었습니다.!

dusdn1702 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,80 @@
+
+public class Attendances {
+
+    private Map<Crew, List<Attendance>> attendances = new HashMap<>();
```
default 값이 아닌 생성자로 생성할 수 있을 것 같아요!

dusdn1702 replied:
List<Attendance> 도 분리해보면 어떨까요?
Attendances는 출석부, List<Attendance>는 크루별 출석정보로 다른 의미를 가질 수 있을 것 같아요.

ljhee92 replied:
그렇네요! 
크루별로 Map을 관리할 생각만 했지 ``List<Attendance>``를 분리할 생각을 못했었어요!
좋은 의견 감사합니다 샐리 반영했습니다!

dusdn1702 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,80 @@
+
+    public Attendance getAttendance(Crew crew, LocalDate localDate) {
+        List<Attendance> attendancesOfCrew = attendances.get(crew);
+        for (Attendance attendance : attendancesOfCrew) {
```
위에서 말씀드린 객체 분리가 된다면 이 부분도 개선이 될 것 같아요!

dusdn1702 commented on src/main/java/attendance/domain/Crew.java:
```diff
@@ -0,0 +1,30 @@
+        return nickName;
+    }
+
+    @Override
+    public final boolean equals(Object o) {
+        if (!(o instanceof Crew crew)) {
+            return false;
+        }
+
+        return Objects.equals(nickName, crew.nickName);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(nickName);
```
요 함수들 override 해주신 이유가 궁금해요~

ljhee92 replied:
같은 닉네임을 가진 크루를 같은 객체로 보기 위해서 override 해두었습니다.

dusdn1702 commented on src/main/java/attendance/domain/Crews.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    public List<Crew> getCrews() {
+        return new ArrayList<>(crews);
```
이렇게 getter 구현해주신 이유가 뭘까요?

ljhee92 replied:
4번 제적 위험자 확인 기능에서 모든 크루별로 제적 위험자인지 확인하기 위해서 꺼냈습니다.
다만 ``Crews``는 필드로 Set을 가지고 있는데 Set은 get으로 요소를 조회할 수 없어서 List로 변환하여 꺼내주었습니다..!

dusdn1702 commented on src/main/java/attendance/domain/Crews.java:
```diff
@@ -0,0 +1,26 @@
+
+public class Crews {
+
+    private Set<Crew> crews = new HashSet<>();
```
여기도 디폴트 값으로 생성하지 않을 수 있을 것 같아요!
디폴트 값으로 객체를 생성했을 때의 단점은 무엇일까요?

ljhee92 replied:
디폴트 값으로 객체를 생성하게 되면 불변을 보장할 수 없어지겠네요..!
객체가 제대로 초기화되고 객체의 상태가 보장되려면 생성자에서 필요한 값들을 주입하는 방식으로 사용하는 방법이 좋을 것 같아요!

dusdn1702 commented on src/main/java/attendance/domain/Penalty.java:
```diff
@@ -0,0 +1,33 @@
+
+    public static Penalty determine(int absenceCount, int lateCount) {
+        absenceCount = absenceCount + lateCount / 3;
+        if (absenceCount > 5) {
```
상수화하면 좋을 것 같아요! 😄

ljhee92 replied:
넵 상수처리 해두었습니다!😄

dusdn1702 commented on src/main/java/attendance/file/AttendanceFileReader.java:
```diff
@@ -0,0 +1,32 @@
+        String str;
+        br.readLine();
+        Attendances attendances = new Attendances();
+        Crews crews = new Crews();
```
file이라는 패키지에서 domain들을 생성하고 있는데요~
IOException이 발생하는데 도메인에 대한 의존도를 낮추면 어떨까요?

ljhee92 replied:
아 이 생각은 하지 못했었네요 ㅠㅠ 도메인에 대한 의존도를 낮출 수 있는 방법이 뭐가 있을지 고민하다가 우선  IOException이 발생하는 파일을 읽어들이는 부분에서는 ``List<String>``으로 읽어 들이고 이 부분에서 IOException이 발생한다면 catch 하고 도메인 생성은 별도로 메서드를 빼서 ``List<String>``을 받아 생성하도록 리팩토링 해보았습니다. 샐리가 생각하시는 다른 방법도 있을까요?

dusdn1702 replied:
저라면 한줄씩 읽어오는 부분과 그 읽어온 값을 통해 객체 생성하는 부분을 분리할 것 같아요. 도메인의 영역과 입출력의 영역을 어디까지 가져가면 좋을지 고민해보아요! 😄

ljhee92 replied:
[refactor: 파일 입출력과 입출력을 통해 읽어온 문자열을 도메인으로 생성하는 로직 분리](https://github.com/ljhee92/java-attendance/commit/79ca73e1acaaa8f83f3c5c338e98862c6c641c80)

파일 입출력도 view와 마찬가지로 문자열을 읽어오는 부분까지만 입출력의 영역으로 보고 반영해보았어요!
그렇다면 파일 입출력도 view와 동일하게 domain과의 상호작용을 controller에서 한다고 봐도 무방할까요?

dusdn1702 commented on src/main/java/attendance/util/DateUtil.java:
```diff
@@ -0,0 +1,20 @@
+
+    private DateUtil() {}
+
+    public static boolean isWeekend(LocalDate date) {
```
같은 이름을 가진 두개의 메서드가 있으면 어떤값을 써야할지 혼란이 오고 실수하기 쉬울 것 같아요.
하나로 합쳐서도 사용할 수 있을 것 같고, 메서드명을 바꿀 수도 있을 것 같네요!

ljhee92 replied:
앗 네 코드 작성 당시에는 시간이 없어 우선 오버로딩하여 사용했었는데, 하나로 합쳐서 사용할 수 있겠네요!

dusdn1702 commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,138 @@
+    private OutputView() {
+    }
+
+    public static void printAttendanceResult(Attendance attendance) {
```
view에서 도메인에 대해 모든 정보를 다 알고 있는데요! view와 domain을 분리해보면 어떨까요?

ljhee92 replied:
MVC 패턴은 view에서는 domain을 알 수 있지만 domain은 view를 몰라야 한다고 알고 있어 domain을 그대로 view에 전달해주도록 구현했는데요!
엄격하게 구분하자면 책임의 분리를 위해 view에서도 domain을 모르는 것이 더 좋다고 생각합니다.
view와 domain을 분리한다고 한다면, domain에서 다른 객체를 생성하여 view로 넘겨주어야 할까요??🤔

dusdn1702 replied:
그렇다면 controller의 역할을 무엇이라고 생각하시나요? 🤔

ljhee92 replied:
이마를 탁 치게 만드는 질문이네요 샐리,,👍
view가 받은 문자열을 controller에서 domain으로 변환해줄 생각은 했으면서 domain이 뱉은 결과값을 controller에서 변환해줄 생각은 못했어요.

[refactor: view에서 domain 분리](https://github.com/ljhee92/java-attendance/commit/236f702b7376107c677e8e0cfe8ac3edae1f3d90)

이렇게 반영해보았습니다. 뷰에서 공통적으로 사용하는 ``AttendanceResult``는 레코드로 변환하여 넘겨주었고, ``PenaltyResult``는 뷰에서 정렬을 책임지게 하고 싶어서 레코드로 변환하여 넘겨주었어요.
그 외 값들은 원시값으로 뷰에게 전달해주었습니다.

dusdn1702 commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,138 @@
+        }
+    }
+
+    public record PenaltyResult(String nickName, int absenceCount, int lateCount,
```
여기서 nested record 선언해주신 것 같은데요~ 
내부 레코드는 외부에서 사용하지 못하도록 한다는 장점이 있을 때 사용했던 것 같아요. 그 외 복잡한 요청응답 매핑이 필요할 때 한군데서 모아보기 위해서도 사용했던 것 같은데요, 내부 레코드로 사용하는데 public으로 선언하면 그 가치가 있을까요?

ljhee92 replied:
아하..! 내부 레코드는 주로 외부에서 사용하지 못하도록 한다는 장점이 있을 때 사용하는지 몰랐었습니다!
만약 내부 레코드로 선언한 객체를 외부에서도 필요로 한다면 따로 레코드 클래스로 생성하는 것이 올바를까요?

dusdn1702 replied:
그럼 외부에서도 사용이 필요한데 내부 레코드로 가져갈 때의 장점이 무엇일까요? 👀

ljhee92 replied:
내부 레코드를 선언한 클래스를 통해서만 내부 레코드에 접근할 수 있고 그 클래스에서 어떠한 이유로 내부 레코드가 선언되었는지 알기 쉬울 것 같아요!

그런데 사실 이 부분도 도메인을 뷰로 꺼내와서 정렬하고 출력하려다 보니 발생한 내부 레코드여서, controller 단에서 출력값을 해결하고 온다면 사용할 필요가 없을 것 같아요!

dusdn1702 commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,138 @@
+
+    private static int countAttendanceStatus(List<Attendance> attendances,
+        AttendanceStatus attendanceStatus) {
+        return Math.toIntExact(
+            attendances.stream()
+                .filter(attendance -> attendance.getStatus() == attendanceStatus)
+                .count()
```
도메인 로직에 가까운 로직이 view에 혼재해있는 것 같아요~
view의 역할에 대해서 질문도 주셨는데요, domain에 대해서 이렇게 view가 의존도가 높으면 안좋은 점이 무엇일까요?
mvc 패턴을 사용하는 이유가 무엇이라고 생각하시는지도 궁금하네요 ㅎㅎ

ljhee92 replied:
이번 미션에서 도메인을 그대로 넘겨주었던 것은 정렬에 대한 책임이 누구에게 있는가에만 초점을 맞추었기 때문인데요! 
'제적 위험자는 제적 대상자, 면담 대상자, 경고 대상자순으로 출력하며, 대상 항목별 정렬 순서는 지각을 결석으로 간주하여 내림차순한다. 출석 상태가 같으면 닉네임으로 오름차순 정렬한다.' 라는 요구 사항을 '출력 요구사항'이라고 생각했기 때문에 뷰에서 카운트하고 정렬을 수행했습니다.
그런데 샐리 말씀처럼 출석 상태라는 도메인에 초점을 맞추게 된다면 출석 상태를 기준으로 정렬하는 것은 도메인의 책임이라고 볼 수 있겠다는 생각이 들었어요.
mvc 패턴을 사용하는 이유는 모델, 컨트롤러, 뷰가 각각 다른 책임을 가지기 위함이고, view가 domain에 대해 의존도가 높다는 것은 변경에 취약하다는 뜻이라고 생각해요.
의존도를 낮추고 변경에 취약하지 않게 하려면 출석 상태를 기준으로 정렬하는 로직을 도메인 내부로 숨기고, 뷰는 '출력'만 담당할 수 있도록 다른 객체라던가 DTO를 던져주는 것도 방법이 될 수 있을 것 같네요!

ljhee92 replied:
'출석 기록 확인'이라는 한 가지 기능을 한 가지 메서드로만 처리하려고 생각했기 때문인 것 같아요.
뷰에서 ``List<Attendance>``를 받아와서 한 번에 처리하려고 하다 보니 도메인 로직에 가까운 로직이 view에 생기게 된 것 같네요.
mvc 패턴을 사용하는 이유는 모델, 컨트롤러, 뷰가 각각 다른 책임을 가지기 위함이고, view가 domain에 대해 의존도가 높다는 것은 변경에 취약하다는 뜻이라고 생각해요.
출석 기록, 상태에 따른 횟수, 패널티 대상이 무엇인지 등을 도메인에서 각각 넘겨주고 뷰는 출력만 담당하게 된다면 책임이 좀 더 명확한 뷰가 될 수 있겠네요!

dusdn1702 commented on src/main/java/attendance/view/OutputView.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    public record PenaltyResult(String nickName, int absenceCount, int lateCount,
+                                Penalty penalty) implements Comparable<PenaltyResult> {
```
Comparable 을 사용하신 이유도 궁금합니다

ljhee92 replied:
'제적 위험자는 제적 대상자, 면담 대상자, 경고 대상자순으로 출력하며, 대상 항목별 정렬 순서는 지각을 결석으로 간주하여 내림차순한다. 출석 상태가 같으면 닉네임으로 오름차순 정렬한다.' 라는 요구사항에 맞게  출력하려는 ``PenaltyResult`` 객체를 정렬하기 위해서 ``Comparable``을 오버라이딩해서 사용했습니다.

dusdn1702 replied:
오버라이딩보다 메서드를 사용해 정렬하면 어떨까요? 오버라이딩으로 얻을 수 있는 장단점을 생각해보면 좋겠어요 👍

ljhee92 replied:
오버라이딩을 하게 되면 여러 곳에서 정렬이 필요할 때 여러 곳에서 재사용할 수 있겠지만 코드가 길어지고 어떻게 정렬하는지 오버라이딩 한 사람 외에는 잘 모를 수 있겠다는 생각이 들어요.
메서드를 사용해 정렬하면 매 번 메서드를 호출해줘야 하지만 가독성이 좋아질 것 같네요!

dusdn1702 commented on src/test/java/attendance/domain/AttendanceTest.java:
```diff
@@ -0,0 +1,93 @@
+
+    @Test
+    void 닉네임과_등교시간으로_출석을_한다() {
+        LocalTime time = LocalTime.of(9, 59);
```
등교시간 규칙이 변경되면 모든 테스트를 다 수정해야할텐데요~!
로직에 변경이 없다면 그 로직을 테스트하는게 단위테스트의 목적이니 변수 수정이 불필요하도록 테스트를 리팩터링해보면 어떨까요?

ljhee92 replied:
음 그럼 테스트 코드에서도 테스트를 위한 constant를 사용해야 할까요?
규칙이 변경되면 모든 테스트를 다 수정해야 할 필요가 없어야 한다는 말씀에는 저도 동의하지만 테스트 코드에서도 사용되는 케이스들을 모두 상수화해야 하는 것인가? 라는 의문이 들어서요..!

dusdn1702 replied:
저라면 기존 프로덕션 코드에서 사용되고 있는 상수를 활용할 것 같아요. 
`LocalTime absentTime = OPERATING_START_TIME.getTime().plusMinutes(30);`
이러한 코드를 사용하면 별다른 상수화 없이 프로덕션 코드에만 의존하는 변수를 만들어낼 수 있지 않을까요?

ljhee92 replied:
[refactor: 테스트에서만 사용되던 getter method 제거](https://github.com/ljhee92/java-attendance/commit/cea464b5f71b1ab7c00ad51464d9ff56f4c5c27e)

``AttendancePolicy`` 객체의 ``getTime()`` 메서드는 테스트에서만 사용되고 있는 메서드인데요!
이런 경우에 샐리가 말해주신 것처럼 ``OPERATING_START_TIME.getTime().plusMinutes(30);``과 같은 테스트 안정성을 위한 메서드 생성은 테스트에서만 사용되어도 괜찮을걸까요?

dusdn1702 commented on src/test/java/attendance/domain/CrewTest.java:
```diff
@@ -0,0 +1,15 @@
+import attendance.domain.Crew;
+import org.junit.jupiter.api.Test;
+
+public class CrewTest {
```
전체적으로 리팩터링 후에 더 많은 단위 테스트가 나올 수 있을 것 같아요 ㅎㅎ
기다리고 있겠습니다 😄

dusdn1702 reviewed (changes requested):
안녕하세요 듀이~
구현해나가는 방식에 있어서 많은 고민을 해주신 것 같아요. 미션 구현하시느라 고생 많으셨습니다 ㅎㅎ

몇가지 궁금한 점과 기능에 있어서 확인 필요한 부분들 리뷰 남겨두었어요.
반영하시면서 궁금한 점은 편하게 댓글 남겨 주시면 같이 확인하도록 하겠습니다 🙇

dusdn1702 commented:
도메인 분리와 범위
‘출석’ 도메인을 먼저 분리하고 출석에서부터 필요한 것들을 분리해나가는 것도 좋은 방법 중 하나인 것 같은데요~! 저라도 말씀 주신 것처럼 우선 돌아가도록 만들고, 거기서 하나의 케이스씩 테스트로 작성하고 분리해가는 과정으로 진행했을 것 같아요 !

공휴일 enum 처리
이건 외부 라이브러리를 가져오지 않고서 지금으로써는 상수로 관리하는 수밖에 없을 것 같아요 ㅜㅜ

view의 책임
오름차순 정렬의 경우 크루라는 객체가 노출을 위해 오름차순 정렬하는 책임까지 가지는 것이 맞을까요? 그렇다면 객체명이 정렬에 대한 내용까지 추가되어야하지 않을까요?

ljhee92 committed:
fix: IndexOutOfBoundsException 해결

sequence가 저장되어 있는 출석/지각 기록의 범위를 넘어갈 때 발생했던 버그를 수정했습니다.

ljhee92 committed:
refactor: datetime 예약어 attendanceTime으로 변경

ljhee92 committed:
refactor: 출석 상태 결정 로직을 Attendance에서 AttendanceStatus로 이동

- 출석 상태 결정 로직을 AttendanceStatus 내부로 이동하였습니다.
- 로직을 이동시키며 setter와 겹치는 메서드명을 수정하였습니다.
- AttendanceStatus와 Penalty의 koreanName을 status로 변경하였습니다.
- 출석 상태를 결정하는 시간, 패널티를 결정하는 숫자의 매직넘버를 상수화했습니다.

ljhee92 committed:
refactor: 오버로딩한 isWeekend 메서드 하나로 통일

ljhee92 committed:
refactor: 출석 규칙 Enum 분리

- 출석 규칙을 한 번에 관리할 수 있도록 AttendancePolicy Enum으로 분리하였습니다.
- 출석인지, 지각인지 확인하는 로직을 AttendancePolicy 내부에서 확인할 수 있게 했습니다.
- 규칙으로 지정된 시간을 AttendanceTest에서도 사용하여 변경을 줄일 수 있도록 했습니다.

ljhee92 committed:
refactor: IOException catch 처리

- 파일 읽기 시 IOException이 발생할 경우 도메인에 주는 영향을 줄이기 위해 메서드를 분리하였습니다.

ljhee92 committed:
refactor: 내부에서만 사용되는 내부 record 접근 제어자 변경

ljhee92 committed:
refactor: 객체 생성 시 default로 생성하지 않고 생성자에 주입

ljhee92 committed:
refactor: List<Attendance> 일급 컬렉션 분리

- 분리 후 1차로 출석 수정 부분만 변경하였습니다.

ljhee92 committed:
refactor: 비즈니스 로직을 도메인 내부로 이동

- 결석 정보를 추가하는 로직을 Attendances 내부로 이동했습니다.
- 상태에 따라 출석, 지각, 결석 횟수를 카운트하는 로직을 AttendanceBook 내부에서 처리하도록 했습니다.

ljhee92 committed:
fix: 출석 수정 시 수정하려는 일자가 등교일인지 검증 추가

ljhee92 committed:
test: Crews, Penalty 테스트 추가

ljhee92 commented:
샐리 안녕하세요, 피드백 반영이 조금 늦었습니다 죄송합니다!

생각보다 생각해야 할 것도 많고 고쳐야 할 부분도 많이 보여서 시간이 오래 걸렸네요🥹
우선 피드백해주신 부분 최대한 반영하려고 노력했고, 리팩토링하면서도 TDD를 지키려고 노력하긴 했지만 철저하게 지키지는 못했습니다.
도메인 중심으로 사고하는 게 아직도 어렵고 익숙하지가 않네요..ㅠㅠ

리뷰해주셔서 항상 감사합니다!☺️

dusdn1702 commented on src/main/java/attendance/domain/AttendancePolicy.java:
```diff
@@ -0,0 +1,48 @@
+        this.time = time;
+    }
+
+    public LocalTime getTime() {
+        return time;
+    }
```
테스트에서만 사용되는 코드를 위해 getter가 존재해야할까요?

ljhee92 replied:
확실히 테스트에서만 사용되는 코드를 위해 getter가 존재할 필요는 없겠네요!
테스트에서 ``getTime()``이 사용되고 있는 부분은 등교날짜가 주말이 아니면 출석 객체가 생성되고, 주말이면 춣석 객체가 생성되지 않고 예외가 터지는지만 테스트하면 되니까 굳이 운영 시간 내의 시간을 주입해주지 않아도 될 것 같아요!

dusdn1702 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,103 @@
+
+    public List<Attendance> getAddAbsenceAttendances(LocalDate today) {
+        int day = today.getDayOfMonth();
+        List<Attendance> copiedAttendances = new ArrayList<>(attendances);
+        copiedAttendances.removeIf(attendance -> attendance.getAttendanceDateTime().getDayOfMonth() == day);
```
이렇게 복사해서 사용하는 것말고 더 좋은 방법을 생각해보면 어떨까요?

ljhee92 replied:
새로운 리스트를 만드는 방법도 있을 것 같아요! 반영해보겠습니다!👍

[refactor: 전날까지의 크루별 출석 기록 확인 시 결석 객체 추가 로직 변경](https://github.com/ljhee92/java-attendance/commit/9df5cb4215e6551feec52a7860d9fe7c6f93536d)

이렇게 변경하였습니다!

dusdn1702 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    private void addAbsenceRecord(int sequence, List<Attendance> copiedAttendancesOfCrew, LocalDate currentDay) {
+        Attendance attendance;
```
이렇게 객체 값을 바꾸는 것은 위험해보여요!

ljhee92 replied:
위 코멘트와 같이 리스트를 복사하지 않고 새로 생성한다면 이 부분도 변경 가능할 것 같아요! 함께 반영해보겠습니다😄

dusdn1702 commented on src/main/java/attendance/domain/Attendances.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    @Override
+    public String toString() {
```
toString은 왜 override 하셨을까요?

ljhee92 replied:
객체의 주소가 아닌 객체가 가지고 있는 값들을 확인해보기 위해 디버깅 용으로 오버라이딩 했습니다!
다만 굳이 남겨 두어야 할 이유가 없어서 삭제처리 했습니다.

[refactor: 전역으로 많이 사용되는 DATE_FORMATTER 변수 DateUtil 클래스 내부로 이동, 디버깅용 toString 제거](https://github.com/ljhee92/java-attendance/commit/a65b96ae2f3e67c48be20d4a4ef321b2f03127fa)

dusdn1702 reviewed (approved):
안녕하세요 듀이!
피드백 반영해주신 부분 확인했습니다 ㅎㅎ 1단계 미션 고생 많으셨어요.
몇가지 코멘트 추가로 남겨두었는데 다음 단계에서 확인 부탁드릴게요!
궁금한 점 있으시면 언제든 편하게 코멘트 남겨주세요 👍
